plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -0,0 +1,570 @@
+package io.jenkins.tools.pluginmanager.impl;
+
+import java.io.File;
+import java.io.InputStream;
+import java.io.FileOutputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.FileWriter;
+import java.io.FileFilter;
+import java.io.RandomAccessFile;
+import java.util.Collections;
+import java.util.List;
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.util.stream.Stream;
+import java.util.Iterator;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.jar.Attributes;
+import java.util.Map;
+import java.util.HashMap;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.Files;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.charset.Charset;
+import java.nio.file.PathMatcher;
+import java.net.MalformedURLException;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.HttpHost;
+import org.apache.http.client.utils.URIUtils;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.HttpStatus;
+
+import org.apache.commons.io.FilenameUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.filefilter.WildcardFileFilter;
+import org.apache.commons.io.IOUtils;
+
+import org.json.JSONObject;
+import org.json.JSONArray;
+
+import io.jenkins.tools.pluginmanager.config.Config;
+
+
+public class PluginManager {
+    private List<Plugin> plugins;
+    private List<Plugin> failedPlugins;
+    private File refDir;
+
+    private String JENKINS_UC_LATEST = "";
+    private final String JENKINS_UC_EXPERIMENTAL = "https://updates.jenkins.io/experimental";
+    private final String JENKINS_INCREMENTALS_REPO_MIRROR = "https://repo.jenkins-ci.org/incrementals";
+    private final String JENKINS_UC = "https://updates.jenkins.io";
+    private final String JENKINS_UC_DOWNLOAD = JENKINS_UC + "/download";
+    private final String JENKINS_UC_JSON = "https://updates.jenkins.io/update-center.json";
```

Comment:
Nit just to make sure that it'll check the correct update center in all the right places:
```suggestion
    private final String JENKINS_UC_JSON = JENKINS_UC + "/update-center.json";
```

Reply:
Resolve in 7ca3d5d987f6c60617f5f95994574849fb17f170.